File: src/twitch/stream_info.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Stream {
    pub id: String,
    pub user_id: String,
    pub user_login: String,
    pub user_name: String,
    pub game_id: String,
    pub game_name: String,
    #[serde(rename = "type")]
    pub stream_type: String,
    pub title: String,
    pub viewer_count: u32,
    pub started_at: String,
    pub language: String,
    pub thumbnail_url: String,
    // D'autres champs peuvent être ajoutés si besoin
}

#[derive(Serialize, Deserialize, Debug)]
struct GetStreamsResponse {
    data: Vec<Stream>,
}

pub async fn get_stream_info(
    broadcaster_id: &str,
    client_id: &str,
    access_token: &str,
) -> Result<Option<Stream>, reqwest::Error> {
    let client = Client::new();
    let url = format!("https://api.twitch.tv/helix/streams?user_id={}", broadcaster_id);

    let response = client
        .get(&url)
        .header("Client-ID", client_id)
        .header("Authorization", format!("Bearer {}", access_token))
        .send()
        .await?;

    // Désérialise la réponse JSON
    let get_streams: GetStreamsResponse = response.json().await?;
    
    // Si le vecteur data n'est pas vide, ça veut dire que l'utilisateur est live.
    Ok(get_streams.data.into_iter().next())
}

File: internal/discord/twitch/stream_info.go
package twitch

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Stream represents the Twitch Helix /streams response for a live stream
type Stream struct {
	ID           string    `json:"id"`
	UserID       string    `json:"user_id"`
	UserName     string    `json:"user_name"`
	GameID       string    `json:"game_id"`
	GameName     string    `json:"game_name"`
	Title        string    `json:"title"`
	ViewerCount  int       `json:"viewer_count"`
	StartedAt    time.Time `json:"started_at"`
	Language     string    `json:"language"`
	ThumbnailURL string    `json:"thumbnail_url"`
	IsMature     bool      `json:"is_mature"`
}

// streamsResponse wraps the JSON response from Twitch Helix
type streamsResponse struct {
	Data []Stream `json:"data"`
}

// GetStreamInfo fetches stream information for the given broadcaster ID.
// Returns a pointer to Stream if live, or nil if offline.
func GetStreamInfo(broadcasterID, clientID, oauthToken string) (*Stream, error) {
	url := fmt.Sprintf("https://api.twitch.tv/helix/streams?user_id=%s", broadcasterID)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Client-ID", clientID)
	req.Header.Set("Authorization", "Bearer "+oauthToken)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode/100 != 2 {
		return nil, fmt.Errorf("twitch API error: %s", resp.Status)
	}

	var sr streamsResponse
	if err := json.NewDecoder(resp.Body).Decode(&sr); err != nil {
		return nil, err
	}

	if len(sr.Data) == 0 {
		return nil, nil // offline
	}
	return &sr.Data[0], nil
}
